evm/src/executor/fork/backend.rs

```diff
@@ -235,100 +238,121 @@ where
+                                    let _ = l.send(value);
+                                })
+                            }
+                            continue
                         }
-                        continue
                     }
-                }
-                ProviderRequest::BlockHash(fut) => {
-                    if let Poll::Ready((block_hash, number)) = fut.poll_unpin(cx) {
-                        let value = block_hash.unwrap_or_else(|_| {
-                            trace!("Failed to get block hash for {}", number);
+                    ProviderRequest::BlockHash(fut) => {
+                        if let Poll::Ready((block_hash, number)) = fut.poll_unpin(cx) {
+                            let value = block_hash.unwrap_or_else(|_| {
+                            trace!( target: "backendhandler", "Failed to get block hash for {}", number);
                             Default::default()
                         });
 
-                        // update the cache
-                        pin.db.block_hashes().write().insert(number, value);
+                            // update the cache
+                            pin.db.block_hashes().write().insert(number, value);
 
-                        // notify all listeners
-                        if let Some(listeners) = pin.block_requests.remove(&number) {
-                            listeners.into_iter().for_each(|l| {
-                                let _ = l.send(value);
-                            })
+                            // notify all listeners
+                            if let Some(listeners) = pin.block_requests.remove(&number) {
+                                listeners.into_iter().for_each(|l| {
+                                    let _ = l.send(value);
+                                })
+                            }
+                            continue
                         }
-                        continue
                     }
                 }
+                // not ready, insert and poll again
+                pin.pending_requests.push(request);
+            }
+
+            // If no new requests have been queued, break to
+            // be polled again later.
+            if pin.queued_requests.is_empty() {
+                break
             }
-            // not ready, insert and poll again
-            pin.pending_requests.push(request);
         }
 
-        // the handler is finished if the request channel was closed and all requests are processed
-        if pin.incoming.is_done() && pin.pending_requests.is_empty() {
-            if let Poll::Ready(Some(ack)) = Pin::new(&mut pin.shutdown).poll_next(cx) {
-                // effectively flushing the cached storage if any
+        if let Poll::Ready(Some(ack)) = Pin::new(&mut pin.shutdown).poll_next(cx) {
+            trace!(target: "backendhandler", "received shutdown signal");
+            if pin.incoming.is_done() {
```

Comment:
there was a problem because we never joined the backgroundhandler to the main thread but added cache flushing to it. which should be triggered if all senders are dropped but we realised we need to and wait on drop until the cache was flushed.